src/rust/ide/view/graph-editor/src/component/visualization/container.rs

```diff
@@ -394,13 +394,16 @@ impl ContainerModel {
         self.set_visibility(!self.is_active())
     }
 
-    fn set_visualization(&self, visualization:Option<visualization::Instance>) {
-        if let Some(visualization) = visualization {
-            let size = self.size.get();
-            visualization.set_size.emit(size);
-            self.view.add_child(&visualization);
-            self.visualization.replace(Some(visualization));
+    fn set_visualization
+    (&self, visualization:visualization::Instance, preprocessor:&frp::Any<EnsoCode>) {
+        let size = self.size.get();
+        visualization.set_size.emit(size);
+        let _vis_network = visualization.network();
+        frp::extend! { _vis_network
+            preprocessor <+ visualization.on_preprocessor_change;
```

Comment:
I think here is a logical bug. If I' not wrong and if I understand this code correctly, when switching a visualization, this code attaches `on_preprocessor_change` output from the new visualization to some outer `preprocessor` FRP input. I see two things here:
1. We do not use in any other file such FRP code that is a function that accepts ref to an `frp` input and attaches things to it. Such logic should rather be expressed in FRP itself because of 2 reasons:
    - It is really hard to track it down.
    - It is easy to make errors this way (see point 2).
2. What happens if we switch visualizations several times? I understand that this will still be bound to old visualizations as well, right? If so, old visualizations can run in parallel and can send `set_preprocessor` events that can then set the preprocessor of other visualization. Even if this is not true, because in our current case old visualizations are destroyed, then I don't like the implicit assumptions of this function - it assumes that the old visualization will be removed before this function is evaluated (if not, then old visualization before being destroyed can still emit `set_preprocessor` event). (Also, we will want to be able to display more than one visualization attached to the same node - however, this may be done as a special case of specially displayed nodes on the stage, so it may not be important here. Still, worth noting.)

Reply:
> We do not use in any other file such FRP code that is a function that accepts ref to an frp input and attaches things to it. 
We do. Look at `new_node` function of `GraphEditorModelWithNetwork`

> What happens if we switch visualizations several times? I understand that this will still be bound to old visualizations as well, right? If so, old visualizations can run in parallel and can send set_preprocessor events that can then set the preprocessor of other visualization. Even if this is not true, because in our current case old visualizations are destroyed, then I don't like the implicit assumptions of this function - it assumes that the old visualization will be removed before this function is evaluated (if not, then old visualization before being destroyed can still emit set_preprocessor event)

It is not an assumption of this function. The old visualization is destroyed in the same place, a line 406. So it will be destroyed before any other code will be run (remember, we are in a single-threaded environment).

And TBH I don't know how to do it better: to have an endpoint in container which emits me the newest preprocessor regardless what instance is currently attached.

Reply:
After all, we decided to make this "connection" as a separate network, managed solely by Visualization Container, so any connections should not leak after instance change.